google/cloud/pubsublite/internal/partition_publisher.cc

```diff
@@ -94,8 +95,14 @@ void PartitionPublisher::Flush() {
 
 future<void> PartitionPublisher::Shutdown() {
   cancel_token_ = nullptr;
-  return service_composite_.Shutdown().then(
-      [this](future<void>) { SatisfyOutstandingMessages(); });
+  future<void> shutdown = service_composite_.Shutdown();
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (reading_) {
+      shutdown = shutdown.then(ChainFuture(reading_->get_future()));
+    }
+  }
+  return shutdown.then([this](future<void>) { SatisfyOutstandingMessages(); });
```

Comment:
Why is it Okay to hold on to `this` in that continuation?  What guarantees the lifetime of the object is long enough?  Do you mind adding a comment explaining it?

Reply:
This could use a comment, but its part of the interface of Service that you wait on the shutdown future before destroying the object. There's nothing enforcing this in the type system however.